File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;


class Solution {
    public int solution(int[] order) {
        int answer = 0;
        int num = 0;
        int idx = 0;
        
        Queue<Integer> conveyer = new LinkedList<>();
        Stack<Integer> stack = new Stack<>();
        
        for(int i = 1; i <= order.length; i++){
            conveyer.add(i);
        }
        
        while(true){
            while(!conveyer.isEmpty()){
                num = conveyer.poll();
                
                if(order[idx] == num){
                    answer++;
                    idx++;
                    
                    while(!stack.isEmpty() && order[idx] == stack.peek()){
                        answer++;
                        idx++;
                        stack.pop();
                    }
                } else{
                    stack.add(num);
                }                                
            }
            
            while(!stack.isEmpty() && order[idx] == stack.peek()){
                answer++;
                idx++;
                stack.pop();
            }
            
            if(conveyer.isEmpty() && 
                ((!stack.isEmpty() && order[idx] != stack.peek()) || (stack.isEmpty() && answer == order.length) ) ){
                break;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.py
from collections import deque


def solution(order):
    N = len(order)
    # 컨테이너 벨트는 큐 자료구조
    container = deque([i for i in range(1, N+1)], N)
    
    
    # 보조 컨테이너 벨트는 스택 자료구조
    stack = []
    
    # 트럭에 실린 짐
    truck = 0
    
    # 트럭이 idx번째 order의 짐을 실어야 함
    order_idx = 0
    
    # 1. order[order_idx] 찾을때까지 컨테이너의 짐을 stack에 삽입
    # 2. 짐 찾으면 트럭에 넣음
    # 3. 스택에서 짐 찾음
    
    # 컨테이너가 비었고, 보조컨테이너가 비지 않았으며 맨 위가 찾는 짐이 아니면 멈춤
    while True:    
        
        while container:
            num = container.popleft()
            if num == order[order_idx]:
                truck += 1
                order_idx += 1
                
                while stack and stack[-1] == order[order_idx]:
                    order_idx += 1
                    truck += 1
                    stack.pop()
            else:
                stack.append(num)
                
        if stack and stack[-1] == order[order_idx]:
            order_idx += 1
            truck += 1
            stack.pop()
                
        
        if (stack and stack[-1] != order[order_idx]) or truck == N:
            break
        
    return truck